### DISCUSSION TITLE
Error in Model development

### DISCUSSION THREAD

**Discussion by son1403** [casact/chainladder-python#296] [Q&A]
I am trying to calculate ldf & cdf to my development. My triangle show as below: 

 `dev = cl.Development(average = "simple").fit(los_2021)`  is  okay but when i run  `dev.ldf_`, i get the error `ValueError: Shape of passed values is (1, 45), indices imply (1, 16)`

![image](https://user-images.githubusercontent.com/87517594/167376515-fc93034e-10ca-4503-a034-2a4cee63ba4a.png)

Could anyone help me with this?  Thank you

**Comment by kennethshsu** [comment]
Hi @son1403,

Can you post the data set here please? Or as an alternative, would you be able to replicate the error using one of the sample datatset included in the package?

**Reply by son1403** [reply]
Because of data privacy policy i cant post the data set here but i can show you few rows. 
![image](https://user-images.githubusercontent.com/87517594/167525964-421a6f51-8331-431a-9d78-575c43935b9e.png)

Then i convert them to triangle by 

![image](https://user-images.githubusercontent.com/87517594/167526005-35a1cf89-5323-47ff-b259-f56b23c8610b.png)
![image](https://user-images.githubusercontent.com/87517594/167526033-6ac41793-1011-4fa8-af6a-cbbf2032d70d.png)
 

I tried cl.Developmentwith genins sample dataset then it works but not with my data :(

**Reply by kennethshsu** [reply]
I think it'll be a bit difficult to help debug this without the dataset.

What version of the chainladder are you on? Can you check by calling `cl.__version__`? The latest is `0.8.12`.

I think your package is outdated, because you would've been warned to specify `cumulative=False` in your initialization of `cl.Triangle(...)`.

**Reply by jbogaardt** [reply]
It may be that the triangle is somehow malformed. It looks like it is OMDM, but doesn't have all diagonals populated well. Prior to any fitting, is the triangle complete and accurate? From the image it looks like there are some development periods with no data. That would cause ldf fits to fail.

**Reply by son1403** [reply]
> I think it'll be a bit difficult to help debug this without the dataset.
> 
> What version of the chainladder are you on? Can you check by calling `cl.__version__`? The latest is `0.8.12`.
> 
> I think your package is outdated, because you would've been warned to specify `cumulative=False` in your initialization of `cl.Triangle(...)`.

I upgraded to `0.8.12` but still the same error

**Reply by son1403** [reply]
> It may be that the triangle is somehow malformed. It looks like it is OMDM, but doesn't have all diagonals populated well. Prior to any fitting, is the triangle complete and accurate? From the image it looks like there are some development periods with no data. That would cause ldf fits to fail.

Yes, it is OMDM
How could i filter development to my triangle.

**Reply by kennethshsu** [reply]
I don't think you need to filter your development out of the triangle, that is not the problem. What @jbogaardt was saying is that your triangle is not complete, you might be missing some valuation periods.

Can you try to convert your triangle to `OQDQ` or `OYDY` by doing `los_2021.gain("OQDQ")` and see if that helps?

Also, can you try to call `los_2021.origin` or `los_2021.development` to see if you are missing anything?

**Reply by son1403** [reply]
I tried to convert my Triangle to OQDQ but it doesnt work. 

![image](https://user-images.githubusercontent.com/87517594/167765540-bc12797d-309b-4cc9-9ad4-05b0dcb3c523.png)

**Reply by kennethshsu** [reply]
Sorry, it should be `los_2021.grain("OQDQ")`!

**Reply by son1403** [reply]
> Sorry, it should be `los_2021.grain("OQDQ")`!

Thank you all @kennethshsu  @jbogaardt. 
Finally it works. 
Appreciate that.

![image](https://user-images.githubusercontent.com/87517594/167766867-9dc07276-a475-43e1-a755-4d9b3309ab56.png)

**Reply by kennethshsu** [reply]
So the problem I think is that like @jbogaardt said, you must be missing some valuation periods. 

You can try running `los_2021.valuation.unique().sort_values()`, it might help you figure out what you are missing:

**Comment by Obalus2022** [comment]
Hey I am also haveing similar challenges:

![image](https://user-images.githubusercontent.com/110035421/181757230-2c10370b-c70c-4ae4-81b0-2672eba7f5ba.png)

 I did check valuations, and looks ok,although I have future valuations as well. dont know if that is the problem

**Comment by jbogaardt** [comment]
@Obalus2022 , this looks odd, but I cannot reproduce. Using the data visible in your screenshot, I get ldfs consistent with the link ratio triangle:
![image](https://user-images.githubusercontent.com/19401133/181776926-5994c74a-5b37-4979-862f-e79ff03f5ffd.png)
Are you on an older version of chainladder (`cl.__version__`)?

On the valuation vector, it does include dates for the unfilled lower half of your triangle, so  the future dates you are seeing is normal.


reprex code:
```python
import chainladder as cl
import pandas as pd
from io import StringIO
print('chainladder version: ', cl.__version__)

data = """
Idx,Acc,Dev,PDLALAE
PMP,2017,2017,449684
PMP,2018,2018,764067
PMP,2019,2019,428453
PMP,2020,2020,1334759
PMP,2021,2021,1316392
PMP,2017,2018,1485157
PMP,2018,2019,1918224
PMP,2019,2020,2357146
PMP,2020,2021,5503980
PMP,2017,2019,2086793
PMP,2018,2020,2042970
PMP,2019,2021,2871223
PMP,2017,2020,2267049
PMP,2018,2021,2081615
PMP,2017,2021,2444177
"""

Triangles = cl.Triangle(
    pd.read_csv(StringIO(data)),
    origin='Acc', development='Dev', 
    index='Idx', columns='PDLALAE',
    cumulative=True)

dev = cl.Development(average='volume').fit(
    Triangles.loc['PMP', 'PDLALAE', :, :]
)
dev.ldf_
```
